# Lesson 5 Assignment
### Jessica Kunkel

As always, the first cells in the Jupyter notebook are setting up the environment.

In [1]:
#!/usr/bin/env python
""" Import display functions and create a function to print outputs to 
Markdown """

from IPython.display import display, Markdown
def printmd(string):
    display(Markdown(string))

In [2]:
""" Import the os module and set the working directory to where the 
mtcars.csv file is saved """

import os
path = r"C:\Users\jkunk\OneDrive\Documents\_SWENG Masters\9 DAAN 862 - Analytics Programming Python"
os.chdir(path)

In [3]:
""" Import NumPy, pandas, and dataframe """
import numpy as np
import pandas as pd
from pandas import DataFrame

### 1. Upload Registration.csv and Course_info.xlsx into Pandas. 

This step is relatively straightforward, using the pandas function `read_csv` and `read_excel` to import the provided data files.

In [4]:
# reg = registration
# course = course_info
reg = pd.read_csv("Registration.csv")
course = pd.read_excel("Course_info.xlsx")

### 2. Explore and clean Registration data.

To explore the Registration data, I first wanted to understand the columns and types of data in each column, as well as the size of the dataset.

In [5]:
reg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4900 entries, 0 to 4899
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Student name  4900 non-null   object
 1   semester new  4900 non-null   object
 2   coursename    4899 non-null   object
dtypes: object(3)
memory usage: 115.0+ KB


The `info` function indicates that there are 4900 entries, and only the coursename column contains a null value.

I then chose to take a sample of the data (indexes 25-34) to see what some of the values looked like. 

In [6]:
reg[25:35]

Student name semester new                  coursename
25          Chris Evert  Spring 2001  Environmental Case Studies
26          Chris Evert  Spring 2002              A WORLD AT WAR
27  Ruth Bader Ginsburg    Fall 2001    AMERICAN SOUTH 1861-PRES
28    Geraldine Ferraro  Summer 2004     BEHAVIORAL PHARMACOLOGY
29    Geraldine Ferraro  Summer 2004     BEHAVIORAL PHARMACOLOGY
30    Geraldine Ferraro  Summer 2004     AMERICAN FOREIGN POLICY
31    Geraldine Ferraro  Summer 2004     AMERICAN FOREIGN POLICY
32    Geraldine Ferraro  Spring 2003       COMPUT LINEAR ALGEBRA
33    Geraldine Ferraro  Spring 2003       COMPUT LINEAR ALGEBRA
34    Geraldine Ferraro  Spring 2003  Environmental Case Studies

From the sample rows, I can see that the information in indexes 28 and 29 are duplicated, as well as indexes 30 and 31, and indexes 32 and 33. I know that I will need to remove duplicates from the dataset.

I then wanted to get a better idea of what values were in the dataset, by each column. I chose to take each column and use the `drop_duplicates` function to retain unique values.

In [7]:
reg["coursename"].drop_duplicates(inplace=False)

0                                 BEHAVIORAL PHARMACOLOGY
1                                 AMERICAN FOREIGN POLICY
2                                   DRUGS, BRAIN AND MIND
3                              Environmental Case Studies
4                                 COMPUTER LINEAR ALGEBRA
                              ...                        
4504                          EURO INT'L REL SINCE WW One
4579    1000 YRS MUSICAL LISTENG: 1000 YRS MUSICAL LIS...
4673                                  DIGITAL PHOTOGRAPHY
4863                             ART AND BUSINESS OF FILM
4894                CREAT.NON-FICTION WRIT: PEER TUTORING
Name: coursename, Length: 169, dtype: object

The coursename column has 169 unique values, but it also looks like some of the courses are capitalized, while others are in title case. This could be another opportunity to clean the dataset.

In [8]:
reg["Student name"].drop_duplicates(inplace=False)

0                 Bill Mumy
24              Chris Evert
27      Ruth Bader Ginsburg
28        Geraldine Ferraro
40              CCH Pounder
               ...         
4861           Gary Crocker
4874               Bob Dole
4885            Stacy Keach
4896            Ann Landers
4898              Tyne Daly
Name: Student name, Length: 448, dtype: object

The Student Name column has unique 448 values, indicating 448 students. The students' names all appear to have appropriate formatting, and will likely not need to be cleaned. 

In [9]:
reg["semester new"].drop_duplicates(inplace=False)

0       Fall 2004
1       Fall 2000
2       Fall 2003
3       Fall 2005
5     Spring 2002
6     Spring 2003
9       Fall 2002
12    Spring 2005
24    Spring 2001
27      Fall 2001
28    Summer 2004
36    Spring 2004
40    Summer 2001
42    Summer 2002
44    Summer 2000
46    Summer 2003
Name: semester new, dtype: object

There are 16 unique semester values. All of the values have appropriate spelling and formatting, and will likely not need to be cleaned.

So then, I got to work cleaning the dataframe. I wanted to know which row or rows contained null values, and used the `isnull` function.

In [10]:
null_indices = reg[reg.isnull().any(axis=1)].index.tolist()
printmd(f"The row containing null values is: {null_indices}")

The row containing null values is: [1650]

Then I created a new dataframe for the cleaned data, called 'reg_cleaned'. I used `dropna` and `drop_duplicates` to get rid of rows that contained null values and rows that were duplicate. For clarity, I decided to reset the index and drop the old index numbering. I cleaned up the column names and their formatting, and sorted the data by student name, in alphabetical order.

In [11]:
# create new data frame that drops null values and duplicates, and resets the index
reg_cleaned = reg.dropna().drop_duplicates().reset_index(drop=True)

# rename the columns of the new dataframe
reg_cleaned.columns = ["Student Name", "Semester", "Course Name"]

# Rename the values in the "Course Name" column to reflect title case
reg_cleaned["Course Name"] = reg_cleaned["Course Name"].str.title()

# ensure all student names are title case
reg_cleaned["Student Name"] = reg_cleaned["Student Name"].str.title()

# Sort alphabetically by student name
reg_cleaned.sort_values(by="Student Name", inplace=True)

# show info on the new, cleaned dataframe
reg_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3650 entries, 3175 to 3165
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Student Name  3650 non-null   object
 1   Semester      3650 non-null   object
 2   Course Name   3650 non-null   object
dtypes: object(3)
memory usage: 114.1+ KB


From the `info` call, you can see that the number of rows has been reduced from 4899 to 3650 by removing duplicate rows and rows containing null values.

In [12]:
reg_cleaned[20:41]

Student Name     Semester  \
2400      Al Gore  Spring 2005   
2395      Al Gore  Spring 2004   
2387      Al Gore  Summer 2000   
2388      Al Gore  Summer 2003   
2362      Al Gore  Summer 2002   
2389      Al Gore  Summer 2001   
2800      Al Hirt    Fall 2003   
2799      Al Hirt    Fall 2002   
2792      Al Hirt    Fall 2001   
2793      Al Hirt    Fall 2000   
2798      Al Hirt    Fall 2004   
2795      Al Hirt    Fall 2002   
2796      Al Hirt    Fall 2004   
2797      Al Hirt    Fall 2001   
2801      Al Hirt    Fall 2003   
2794      Al Hirt    Fall 2000   
2711     Al Roker  Spring 2005   
2708     Al Roker  Spring 2004   
2709     Al Roker  Summer 2001   
2707     Al Roker  Summer 2004   
2710     Al Roker  Summer 2000   

                                            Course Name  
2400  20Th Century Russian Literature: Fiction And R...  
2395                               Analytical Mechanics  
2387                                 Britain Since 1945  
2388  Being Human: Being Human: Biology, Culture & H...  
2362                            American Foreign Policy  
2389                              American Healt Policy  
2800                             Cell Biology & Biochem  
2799                           Food/Feast Arch Of Table  
2792                            Behavioral Pharmacology  
2793                           Environmental Systems Ii  
2798        Dna, Diet, And Disease: The Biology Of Food  
2795                               Analytical Mechanics  
2796                           Contemporary African Art  
2797                                     A World At War  
2801                                Elementary French I  
2794                            Computer Linear Algebra  
2711                               Analytical Mechanics  
2708                              Comput Linear Algebra  
2709                              American Healt Policy  
2707                           Food/Feast Arch Of Table  
2710                                 Britain Since 1945

From the sample above you can see:
- There are no duplicate rows in which the same student, in the same semester, is taking the same class more than once
- The course names are all in title-case, rather than a mix of all capital letters and other cases
- The indexes are reset such that there are no skipped index values due to the duplicate rows removed
- The columns are renamed for clarity and consistency with the Course Info data

### 3. Explore and clean Course info data.

To explore the Course Info data, I first wanted to understand the columns and types of data in each column, as well as the size of the dataset.

In [13]:
course.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42 entries, 0 to 41
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Course number  42 non-null     object
 1   Course Name    41 non-null     object
 2   Course Type    42 non-null     object
dtypes: object(3)
memory usage: 1.1+ KB


The `info` function indicates that therre are 42 rows in the Course Info dataset, and there is one null value in the Course Name column.

I then used the `head` function to get a glimpse of the data.

In [14]:
course.head()

Course number                                     Course Name  Course Type
0       ARTS400  EXPERIMENTAL WRITING SEM: The Ecology of Poetry           C
1       ARTS401                             ART: ancient to 1945           C
2       ARTS465                         ENVIRONMENTAL SYSTEMS II           F
3       ARTS486                          COMPUTER LINEAR ALGEBRA           F
4       ARTS512                             ANALYTICAL MECHANICS           F

From the first 5 rows of data, I can see that the Course Name column appears to have the same inconsistent capitalization scheme as in the Registration dataset. There don't appear to be any duplicate values for the Course Number and Course Name columns, but I decided to go ahead and look at the unique values for each column.

In [15]:
course["Course number"].drop_duplicates(inplace=False)

0      ARTS400
1      ARTS401
2      ARTS465
3      ARTS486
4      ARTS512
5      ARTS514
6      ARTS516
7      ARTS518
8      ARTS520
9     ARTS488 
10     ARTS541
11    ARTS543 
12     ARTS491
13    ARTS492 
14     ARTS493
15     ARTS494
16    ARTS495 
17     ARTS496
18     ARTS497
19    ARTS545 
20     ARTS547
21     ARTS549
22     ARTS551
23    ARTS553 
24     ARTS555
25     ARTS485
26     ARTS484
27    ARTS557 
28    ARTS559 
29     ARTS561
30     ARTS565
31    ARTS567 
32    ARTS569 
33    ARTS571 
34     ARTS573
35     ARTS575
36     ARTS577
37     ARTS579
38    ARTS581 
39    ARTS583 
40    ARTS585 
41     ARTS587
Name: Course number, dtype: object

It appears as though there are 42 unique course numbers, matching the number of rows in the initial dataset. So there are not any duplicate course number values. 

I want to note here that I was having difficulty isolating the Course Name column due to an index error. So I decided to call the `columns` attribute of the course dataframe.

In [16]:
course.columns

Index(['Course number', 'Course Name ', 'Course Type'], dtype='object')

This showed that there was no index 'Course Name' because there was a trailing white space, making the index 'Course Name '. I was then able to use the correct index to get the unique values for the course names to explore this column of the data.

In [17]:
course["Course Name "].drop_duplicates(inplace=False)

0       EXPERIMENTAL WRITING SEM: The Ecology of Poetry
1                                  ART: ancient to 1945
2                              ENVIRONMENTAL SYSTEMS II
3                               COMPUTER LINEAR ALGEBRA
4                                  ANALYTICAL MECHANICS
5                                        A WORLD AT WAR
6                               BEHAVIORAL PHARMACOLOGY
7                              CONTEMPORARY AFRICAN ART
8                              FOOD/FEAST ARCH OF TABLE
9                                 DEVIL'S PACT LIT/FILM
10                               AMERICAN SOCIAL POLICY
11                                     ART AND RELIGION
12                             CONTEMPORARY POL.THOUGHT
13        AFRICAN-AMERICAN LIT: AFRICAN-AMER LIT:CHANGE
14                               AMERICAN HEALTH POLICY
15                 Business German: A Micro Perspective
16                             COMM and  THE PRESIDENCY
17                             French Thought Ti

From the above printout I can see there are 42 unique course name values, and one is null. So that row will need to be removed.

In [18]:
course["Course Type"].drop_duplicates(inplace=False)

0    C
2    F
9    E
Name: Course Type, dtype: object

There are 3 possible course types, but I'm not worried about there being duplicates in this column, as it does not seem like the data in this column needs to be unique. I.e., multiple courses can be of the same type.

Then, I moved on to cleaning the data. Like before, I created a new dataframe for the cleaned data, 'course_cleaned'. I only used `dropna` since there were no notable duplicated rows in the dataset. I renamed and reformatted the columns for consistency, and sorted the dataset by student name in alphabetical order.

In [19]:
# create new dataframe that drops rows with null values and resets the indexes
course_cleaned = course.dropna().reset_index(drop=True)

# rename the columns
course_cleaned.columns = ["Course Number", "Course Name", "Course Type"]

# Rename the values in the "Course Name" column to reflect title case
course_cleaned["Course Name"] = course_cleaned["Course Name"].str.title()

# sort alphabetically by student name
course_cleaned.sort_values(by="Course Number", inplace=True)

# show info on the new, cleaned data
course_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 41 entries, 0 to 40
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Course Number  41 non-null     object
 1   Course Name    41 non-null     object
 2   Course Type    41 non-null     object
dtypes: object(3)
memory usage: 1.3+ KB


From the `info` call, you can see that the number of rows has been reduced from 42 to 41 by removing the row containing a null value.

In [20]:
course_cleaned.head()

Course Number                                      Course Name Course Type
0        ARTS400  Experimental Writing Sem: The Ecology Of Poetry           C
1        ARTS401                             Art: Ancient To 1945           C
2        ARTS465                         Environmental Systems Ii           F
26       ARTS484                          Europe In A Wider World           E
25       ARTS485         Evidenced Based Crime And Justice Policy           E

From the sample above you can see:
- The course names are all in title-case, rather than a mix of all capital letters and other cases
- The indexes are reset such that there are no skipped index values due to the row containing a null value being removed
- The columns are renamed to reduce index errors

### 4. Which course has the highest registration?

To retrieve the course with the highest registration, I chose to use the `value_counts` function on the Course Name column in the cleaned registration information dataset. The `value_counts` function returns a series with Course Name as the index, and the count as the value, in descending order. I was then able to take the first index and the first element and return those to give the course with the highest registration.

In [21]:
counts = reg_cleaned["Course Name"].value_counts()

printmd(f"The course with the highest registration is {counts.index[0]} with {counts.iloc[0]} registrations.")

The course with the highest registration is Comput Linear Algebra with 303 registrations.

### 5. Inner join two datasets.

To inner join the two datasets, it made the most sense to me to first organize by the course number, then by the students enrolled. So I used the `pd.merge` function, and used `course_cleaned` as the first dataset and `reg_cleaned` as the second dataset. Just to be certain, I joined the datasets on the "Course Name" index.

In [30]:
joined = pd.merge(
    course_cleaned, 
    reg_cleaned, 
    on="Course Name", 
    how="inner"
)
joined

Course Number                                      Course Name  \
0          ARTS400  Experimental Writing Sem: The Ecology Of Poetry   
1          ARTS400  Experimental Writing Sem: The Ecology Of Poetry   
2          ARTS400  Experimental Writing Sem: The Ecology Of Poetry   
3          ARTS400  Experimental Writing Sem: The Ecology Of Poetry   
4          ARTS400  Experimental Writing Sem: The Ecology Of Poetry   
...            ...                                              ...   
1745       ARTS587                             Elementary Arabic Ii   
1746       ARTS587                             Elementary Arabic Ii   
1747       ARTS587                             Elementary Arabic Ii   
1748       ARTS587                             Elementary Arabic Ii   
1749       ARTS587                             Elementary Arabic Ii   

     Course Type        Student Name     Semester  
0              C            Al Roker    Fall 2005  
1              C         Ali Macgraw    Fall 2001  
2              C   Annette Funicello    Fall 2003  
3              C  Barbara Bel Geddes  Summer 2004  
4              C    Barbara Delinsky    Fall 2001  
...          ...                 ...          ...  
1745           E         Sally Field  Summer 2002  
1746           E         Soupy Sales  Spring 2001  
1747           E         Steve Berry  Spring 2004  
1748           E         Tip O'Neill  Spring 2004  
1749           E       Wolfgang Puck  Summer 2004  

[1750 rows x 5 columns]

Based on the fact that there are only 1,750 rows in the merged dataset, it seems like there are a fair number of courses in the Registration Info dataset that do not have a corresponding course and course number in the Course Info dataset.

### 6. Create a data frame with student names as the index, course numbers as columns, and if the student registered a course as values (0, 1).

To accomplish this, I used a pivot table that aggregates the max value and returns a 1, to handle cases where a student registers for the same class in more than one semester. The rest of the cell values (i.e., NaNs) are left as zeroes. I chose to create a new dataframe for this operation, called 'joined2', so that I could add a flag column called 'Registration'.

In [34]:
joined2 = joined

# Add a helper column to flag registration
joined2['Registered'] = 1

# Pivot to get 1 where registration exists
result = pd.pivot_table(
    merged_df,
    index='Student Name',
    columns='Course Number',
    values='Registered',
    aggfunc='max',      # We just care whether it exists (1)
    fill_value=0        # Fill NaNs with 0
).astype(int)           # Ensure values are int (0 or 1)

result

Course Number   ARTS400  ARTS401  ARTS465  ARTS484  ARTS485  ARTS486  \
Student Name                                                           
Abella Abzug          0        0        0        0        0        0   
Al Gore               0        0        0        0        0        0   
Al Hirt               0        0        1        0        0        1   
Al Roker              1        0        0        0        0        0   
Alan Bates            0        0        0        0        0        0   
...                 ...      ...      ...      ...      ...      ...   
Willis Johnson        1        0        0        0        0        0   
Winona Ryder          0        0        0        0        0        0   
Wolfgang Puck         0        0        0        0        0        0   
Yogi Berra            0        0        0        0        0        0   
Yoko Ono              0        0        0        0        0        0   

Course Number   ARTS488   ARTS491  ARTS492   ARTS493  ...  ARTS555  ARTS559   \
Student Name                                          ...                      
Abella Abzug           0        0         0        0  ...        0         0   
Al Gore                0        0         0        0  ...        0         0   
Al Hirt                0        0         0        0  ...        0         0   
Al Roker               0        0         0        0  ...        0         0   
Alan Bates             0        0         0        0  ...        1         0   
...                  ...      ...       ...      ...  ...      ...       ...   
Willis Johnson         0        0         0        0  ...        0         1   
Winona Ryder           0        0         0        0  ...        0         0   
Wolfgang Puck          0        0         0        1  ...        0         0   
Yogi Berra             0        0         0        0  ...        0         0   
Yoko Ono               0        0         0        0  ...        0         0   

Course Number   ARTS565  ARTS569   ARTS571   ARTS573  ARTS577  ARTS581   \
Student Name                                                              
Abella Abzug          0         0         0        0        0         1   
Al Gore               0         0         0        0        0         0   
Al Hirt               0         0         0        0        0         0   
Al Roker              0         0         0        0        0         0   
Alan Bates            0         0         0        0        0         0   
...                 ...       ...       ...      ...      ...       ...   
Willis Johnson        0         0         0        0        0         0   
Winona Ryder          0         1         0        0        0         0   
Wolfgang Puck         0         0         0        0        0         0   
Yogi Berra            0         0         0        0        0         0   
Yoko Ono              0         0         0        0        0         0   

Course Number   ARTS583   ARTS587  
Student Name                       
Abella Abzug           0        0  
Al Gore                0        0  
Al Hirt                0        0  
Al Roker               0        0  
Alan Bates             0        0  
...                  ...      ...  
Willis Johnson         0        0  
Winona Ryder           0        0  
Wolfgang Puck          0        1  
Yogi Berra             0        0  
Yoko Ono               0        0  

[408 rows x 34 columns]

Above is the printout of the final dataframe, with student names as the index, course numbers as columns, and if the student registered a course as values (0, 1).